Fixes for Carousel Logical Tree

danwalmsley opened the pull request.

Commit:
Added failing unit test for Carousel

Commit:
initial attempt to set logical children correctly.

Commit:
Merge branch 'master' into fixes/failing-unit-test

Commit:
Items control adds the content presenters child as its logical child.

Commit:
Items control removes the logical child correctly.

Commit:
Updated ItemsControl test for new behaviour.

Commit:
Add ContainerType property to IItemContainerGenerator

Commit:
Add implementation of ContainerType property

Commit:
ItemsControl uses the ContainerType property to check if its a typed container.

Commit:
Fix ItemsControl tests.

grokys commented on src/Avalonia.Controls/Generators/ItemContainerGenerator.cs:
```diff
@@ -50,6 +50,14 @@ public ItemContainerGenerator(IControl owner)
+                return null;
+            }
+        }
+
```
This could be written more succinctly as:

```
public virtual Type ContainerType => null;
```

grokys commented on src/Avalonia.Controls/Generators/ItemContainerGenerator`1.cs:
```diff
@@ -34,6 +34,15 @@ public class ItemContainerGenerator<T> : ItemContainerGenerator where T : class,
+                return typeof(T);
+            }
+        }
+
```
Again here.

grokys commented on src/Avalonia.Controls/ItemsControl.cs:
```diff
@@ -229,12 +232,20 @@ protected virtual void OnContainersMaterialized(ItemContainerEventArgs e)
 
             foreach (var container in e.Containers)
-            {
+            {                
```
Whitespace

grokys commented on src/Avalonia.Controls/ItemsControl.cs:
```diff
@@ -229,12 +232,20 @@ protected virtual void OnContainersMaterialized(ItemContainerEventArgs e)
+                    if (ItemContainerGenerator.ContainerType == null)
+                    {
+                        (container.ContainerControl as ContentPresenter).UpdateChild();
+                        toAdd.Add((container.ContainerControl as ContentPresenter).Child as ILogical);
```
Cast this only once. Also if you're doing an `as` you need to check for null.

grokys commented on src/Avalonia.Controls/ItemsControl.cs:
```diff
@@ -256,7 +267,14 @@ protected virtual void OnContainersDematerialized(ItemContainerEventArgs e)
-                    toRemove.Add(container.ContainerControl);
+                    if (ItemContainerGenerator.ContainerType == null)
+                    {
+                        toRemove.Add((container.ContainerControl as ContentPresenter).Child);
```
Again, either use a standard cast here or if we want to handle cases where the item isn't a `ContentPresenter` put in a null check.

Commit:
fix whitespace

Commit:
ContainerType property implementations using improved syntax.

Commit:
removed excessive casting.

Commit:
whitespace

danwalmsley commented on src/Avalonia.Controls/ItemsControl.cs:
```diff
@@ -219,6 +221,7 @@ protected virtual IItemContainerGenerator CreateItemContainerGenerator()
             return new ItemContainerGenerator(this);
         }
 
+
```
done

Commit:
ItemsControl ensure Containers Templated parent is null.

Commit:
Fix ItemsControl preventing DataTemplates getting applied.

Commit:
fix bug checking wrong property for is ILogical

Commit:
Null check in ItemsControl on Materialized and Dematerialized, to prevent 

exceptions being thrown, trying to do is check on null.

Commit:
Corrected logic for materialize methods.

Commit:
remove excessive null check, is also serves for null check.

Commit:
Set TemplatedParentProperty null before calling update child

so it doesnt end up seting the ItemsControls parent as the container.childs parent.
Preventing it being removed from logical tree.

danwalmsley commented:
@grokys last commit on this fixed a major bug for me. Basically if I didn't do the SetTemplatedParent null
before calling update child, then the Child would end up with the ItemsControls templatedParent as its Parent, this prevented the item getting removed from logical tree.

I spent ages really debugging this stuff in detail.. I would like to add a unit test to check this too.

I'm wondering if because of this we should remove this line...
https://github.com/VitalElement/AvalonStudio/commit/7b75d5a5a5a7332c3aa2dd3cdf669b7191fc384d

with the blog post you did on logical tree weirdness, then something could be a logical child and you want to remove it, even though its logical parent the control that wants to remove it... make sense?

Commit:
Removed comments

Commit:
Added unit test to assert new behavior.

Commit:
whitespace

grokys commented on src/Avalonia.Controls/ItemsControl.cs:
```diff
@@ -219,26 +221,43 @@ protected virtual IItemContainerGenerator CreateItemContainerGenerator()
+
+                            if (containerControl.Child is ILogical)
+                            {
+                                LogicalChildren.Add(containerControl.Child as ILogical);
```
You've already checked if `containerControl.Child` is an `ILogical` above, so this is unnecessarily casting twice.

grokys commented on src/Avalonia.Controls/ItemsControl.cs:
```diff
@@ -256,7 +275,20 @@ protected virtual void OnContainersDematerialized(ItemContainerEventArgs e)
+
+                        if (containerControl != null && containerControl.Child is ILogical)
+                        {
+                            toRemove.Add(containerControl.Child as ILogical);
```
Again, casting twice here.

grokys commented on tests/Avalonia.Controls.UnitTests/ItemsControlTests.cs:
```diff
@@ -62,6 +64,32 @@ public void Container_Should_Have_TemplatedParent_Set_To_Null()
+
+                var container = (ContentPresenter)target.Presenter.Panel.Children[0];
+
+                Assert.Equal(container.Child.Parent, container);
```
The order for `Assert.Equals` arguments is `expected, actual` so these two args should be swapped.

grokys commented:
LGTM, just added a few requests for small changes. Also your previous comment said:

> I'm wondering if because of this we should remove this line...
> VitalElement/AvalonStudio@7b75d5a

Is that the link you meant to paste?

danwalmsley commented:
I think this was the link I wanted to paste...

https://github.com/AvaloniaUI/Avalonia/blob/dec5682dce412f93b3f3bd85c4a697c726e4af78/src/Avalonia.Controls/Control.cs#L781

My thoughts were that it should always clear parent here, because it would prevent from being removed from logical tree, it the i.LogicalParent != this, which could happen?

But now thinking about it, seems like that would not be good!

Commit:
Add missing inheritdoc comment.

Commit:
Removed whitespace.

Commit:
Removed casting that was not needed.

Commit:
Corrected order of Assert arguments.

Commit:
more whitespace changes.

grokys commented:
Yeah, I think the current behavior is correct. So PR LGTM (Lets Get This Merged) =D

Commit:
Removed unused using.

grokys commented on src/Avalonia.Controls/ItemsControl.cs:
```diff
@@ -226,19 +225,35 @@ protected virtual IItemContainerGenerator CreateItemContainerGenerator()
+
+                            containerControl.UpdateChild();
+
+                            if (containerControl.Child is ILogical)
```
Only just noticed now that `containerControl.Child` is `IControl` which inherits from `ILogical` so this check is unnecessary.

grokys commented on src/Avalonia.Controls/ItemsControl.cs:
```diff
@@ -256,7 +271,20 @@ protected virtual void OnContainersDematerialized(ItemContainerEventArgs e)
+                    {
+                        var containerControl = container.ContainerControl as ContentPresenter;
+
+                        if (containerControl != null && containerControl.Child is ILogical)
```
Also this `is` check is unnecessary.

danwalmsley commented on src/Avalonia.Controls/ItemsControl.cs:
```diff
@@ -226,19 +225,35 @@ protected virtual IItemContainerGenerator CreateItemContainerGenerator()
+
+                            if (containerControl.Child is ILogical)
+                            {
+                                LogicalChildren.Add(containerControl.Child);
```
@grokys  Is it more efficient to add them to a list first, then do addrange.

I had at some point during the development needed to add them as I went along for some reason.
might be able to change back if there is a performance hit?

danwalmsley replied:
@grokys this one...

Commit:
No longer type checking as not necessary.

Commit:
Remove allocation of list that is not neccesary

Commit:
Fix null check around adding Container.Child to logical children.

grokys commented:
Thanks Dan! Nice work :)
